visualization/app/codeCharta/state/effects/focusNodes/focus.effect.ts

```diff
@@ -0,0 +1,64 @@
+import { Injectable } from "@angular/core"
+import { Actions, createEffect, ofType } from "@ngrx/effects"
+import { pairwise, startWith, tap, withLatestFrom } from "rxjs"
+import { focusNode, unfocusAllNodes, unfocusNode } from "../../store/dynamicSettings/focusedNodePath/focusedNodePath.actions"
+import { ThreeMapControlsService } from "../../../ui/codeMap/threeViewer/threeMapControls.service"
+import { Store } from "@ngrx/store"
+import { CcState } from "../../../codeCharta.model"
+import { focusedNodePathSelector } from "../../store/dynamicSettings/focusedNodePath/focusedNodePath.selector"
+import { currentFocusedNodePathSelector } from "../../store/dynamicSettings/focusedNodePath/currentFocused.selector"
+import { isChildPath } from "../../../util/isChildPath"
+
+@Injectable()
+export class FocusEffects {
+    constructor(
+        private actions$: Actions,
+        private store: Store<CcState>,
+        private threeMapControlsService: ThreeMapControlsService
+    ) {}
+
+    focusNode$ = createEffect(
+        () =>
+            this.actions$.pipe(
+                ofType(focusNode),
+                withLatestFrom(this.store.select(currentFocusedNodePathSelector).pipe(startWith(null), pairwise())),
+                tap(([, [previousFocusedNodePath, newFocusedNodePath]]) => {
+                    if (previousFocusedNodePath && !isChildPath(newFocusedNodePath, previousFocusedNodePath)) {
+                        this.threeMapControlsService.unfocusNode(() => {
+                            this.threeMapControlsService.focusNode(newFocusedNodePath)
+                        })
+                    } else {
+                        this.threeMapControlsService.focusNode(newFocusedNodePath)
```

Comment:
This action is performed both inside the if and in the else part of the condition. Can it be moved outside or is that not pissible because its used as a callback in unfocusNode?

Reply:
Within the if statement, the callback function is invoked to allow the chaining of the unfocus and focus animations. Without this callback mechanism, only the focus action would need to be triggered directly.